### PR TITLE
Refactor AD

### PR DESCRIPTION
**Pull request by michel2323**

### CODE REVIEW

File: src/Polar/derivatives.jl [hunk 1]
```diff
@@ -83,46 +74,58 @@ No allocations are taking place in this function.
 * `buffer::PolarNetworkState`: store current values for network's variables.
 
 """
-function AutoDiff.jacobian!(polar::PolarForm, jac::AutoDiff.Jacobian, buffer)
+function AutoDiff.jacobian!(
+    polar::PolarForm,
+    jac::AutoDiff.Jacobian{Func, VI, VT, MT, SMT, VP, VD, SubT, SubD, State},
```
Comment (frapac) [inline src/Polar/derivatives.jl:79]: Interesting!

**Review by frapac — APPROVED** [review]
Looks good to me! 
I am ok to merge once the issue with the naming (CamelCase vs under_score) has been addressed :)

File: src/Polar/batch.jl [hunk 2]
```diff
@@ -126,25 +126,14 @@ function BatchJacobian(
...
     device = polar.device
-
-    if isa(device, CPU)
-        SMT = SparseMatrixCSC{Float64,Int}
-        A = Array
-    elseif isa(device, GPU)
-        SMT = CUSPARSE.CuSparseMatrixCSR{Float64}
-        A = CUDA.CuArray
-    end
+    (SMT, A) = get_BatchJacobian_types(device)
```
[the remark this answers is not in the document]
Comment (michel2323) [reply src/Polar/batch.jl]: Fixed.

File: src/Polar/batch.jl [hunk 3]
```diff
@@ -126,25 +126,14 @@ function BatchJacobian(
...
     TT = A{T, 3}
 
     pf = polar.network
     nbus = PS.get(pf, PS.NumberOfBuses())
-    if isa(variable, State)
-        map = VI(polar.mapx)
-    elseif isa(variable, Control)
-        map = VI(polar.mapu)
-    end
+    map = VI(get_map(polar, variable))
```
[the remark this answers is not in the document]
Comment (michel2323) [reply src/Polar/batch.jl]: Fixed.

File: src/Polar/derivatives.jl [hunk 4]
```diff
@@ -18,22 +20,11 @@ function AutoDiff.Jacobian(
...
     @assert is_constraint(func)
-
-    if isa(polar.device, CPU)
-        SMT = SparseMatrixCSC{Float64,Int}
-        A = Vector
-    elseif isa(polar.device, GPU)
-        SMT = CUSPARSE.CuSparseMatrixCSR{Float64}
-        A = CUDA.CuVector
-    end
+    (SMT, A) = get_Jacobian_types(polar.device)
```
[the remark this answers is not in the document]
Comment (michel2323) [reply src/Polar/derivatives.jl]: Fixed.

**Pull request merged**